subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/compile/CrossCompilationIntegrationTest.groovy

```diff
@@ -0,0 +1,128 @@
+import org.junit.Assume
+import org.objectweb.asm.Opcodes
+import spock.lang.Unroll
+
+class CrossCompilationIntegrationTest extends AbstractIntegrationSpec {
+    @Unroll
+    @ToBeFixedForInstantExecution
+    def "can configure the Java plugin to compile and run tests against Java #version JDK"() {
+        def jvm = AvailableJavaHomes.getJdk(version)
+        Assume.assumeTrue(jvm != null)
+
+        buildFile << """
+            plugins {
+                id "java"
+            }
+            repositories {
+                jcenter()
+            }
+            dependencies {
+                testImplementation("junit:junit:4.12")
+            }
+            java {
+                sourceCompatibility = JavaVersion.${version.name()}
+            }
+            
+            def javaInstallation = project.javaInstalls.installationForDirectory(file("${jvm.javaHome.toURI()}"))
+
+            tasks.named("compileJava") {
+                options.fork = true
+                options.forkOptions.javaHome = javaInstallation.get().installationDirectory
+            }
+            tasks.named("compileTestJava") {
+                options.fork = true
+                options.forkOptions.javaHome = javaInstallation.get().installationDirectory
+            }
+            tasks.named("test") {
+                executable = javaInstallation.get().javaExecutable
+            }
+        """
+
+        def sourceFile = file("src/main/java/Thing.java")
+        sourceFile << """
+            import java.util.List;
+            import java.util.ArrayList;
+            
+            public class Thing {
+                public Thing() {
+                    // Uses Java 8 features
+                    Runnable r = () -> { };
+                    r.run();
+                }
+            }
+        """
+
+        def testFile = file("src/test/java/ThingTest.java")
+        testFile << """
+            import org.junit.Test;
+            import org.junit.Assert;
+            import java.io.File;
+            
+            public class ThingTest {
+                @Test
+                public void check() {
+                    Assert.assertTrue(new File(System.getProperty("java.home")).toURI().toString().startsWith("${jvm.javaHome.toURI()}"));
+                    Assert.assertTrue(System.getProperty("java.version").startsWith("${version}."));
+                    new Thing();
+                }
+            }
+        """
+
+        when:
+        fails("build")
+
+        then:
+        failure.assertHasDescription("Execution failed for task ':compileJava'.")
+        failure.assertHasCause("Compilation failed with exit code")
+
+        when:
+        sourceFile.replace("() -> { }", "new Runnable() { public void run() { } }")
+        run("build", "-i")
+
+        then:
+        versionFromByteCode(file("build/classes/java/main/Thing.class")) == version
+        versionFromByteCode(file("build/classes/java/test/ThingTest.class")) == version
```

Comment:
If you're building the jar anyways, you can get the class file version with an existing fixture like in https://github.com/gradle/gradle/blob/e381200cb64d99809e6f8c02c224a1b9e9b46773/subprojects/language-java/src/integTest/groovy/org/gradle/language/java/JavaLanguageIntegrationTest.groovy#L71

Reply:
thanks. I knew we had something to do this but I couldn't find it